src-manual/Vulkan/Requirements.hs

```diff
@@ -184,162 +237,246 @@ data Requirements where
-       , missingFeatures    :: PhysicalDeviceFeatures2   feats -> HashSet ByteString
+       --
+       -- Argument: optional settings that the function will __not__ attempt to enable.
+       , enableSettingsOtherThan :: HashSet ByteString -> Chain setts
+       -- | Returns all the required and optional features that were not enabled in 'PhysicalDeviceFeatures2'.
+       , missingFeatures         :: PhysicalDeviceFeatures2   feats -> ReqAndOpt (HashSet ByteString)
        -- | All required features, to be used in 'DeviceCreateInfo'.
-       , neededFeatures     :: PhysicalDeviceFeatures2   feats
-       -- | Returns all the required properties that were not satisfied in 'PhysicalDeviceProperties2'
-       , missingProperties  :: PhysicalDeviceProperties2 props -> HashSet ByteString
+       --
+       -- Argument: optional features that should __not__ be enabled.
+       -- These must be the missing optional features returned by 'missingFeatures'.
+       , enableFeaturesOtherThan :: HashSet ByteString -> PhysicalDeviceFeatures2 feats
+       -- | Returns all the required and optional properties that were not satisfied in 'PhysicalDeviceProperties2'
+       , missingProperties       :: PhysicalDeviceProperties2 props -> ReqAndOpt (HashSet ByteString)
        }
     -> Requirements
 
 -- | Collect up requirements.
-requirements :: Foldable f => f Requirement -> Requirements
+requirements :: Foldable f => f OptionalRequirement -> Requirements
 requirements = foldr' addRequirement noRequire
   where
     noRequire :: Requirements
     noRequire =
       Requirements @'[] @'[] @'[]
-        (MAKE_VERSION 1 0 0)
-        HashMap.empty HashMap.empty HashMap.empty
-        ()
-        (const HashSet.empty) zero (const HashSet.empty)
-    addRequirement :: Requirement -> Requirements -> Requirements
-    addRequirement (RequireVersion {..}) =
-      requireVersion version
-    addRequirement (RequireInstanceLayer {..}) =
-      requireInstanceLayer instanceLayerName instanceLayerMinVersion
-    addRequirement (RequireInstanceExtension {..}) =
-      requireInstanceExtension instanceExtensionLayerName instanceExtensionName instanceExtensionMinVersion
-    addRequirement (RequireDeviceExtension {..}) =
-      requireDeviceExtension deviceExtensionLayerName deviceExtensionName deviceExtensionMinVersion
-    addRequirement (RequireInstanceSetting {..}) =
-      requireInstanceSetting setInstanceSetting
-    addRequirement (RequireFeature {..}) =
-      requireFeature featureName checkFeature enableFeature
-    addRequirement (RequireProperty {..}) =
-      requireProperty propertyName checkProperty
+        (ReqAndOpt (MAKE_VERSION 1 0 0) (MAKE_VERSION 1 0 0))
+        (ReqAndOpt HashMap.empty HashMap.empty)
+        HashMap.empty
+        HashMap.empty
+        (const ())
+        (const $ ReqAndOpt HashSet.empty HashSet.empty)
+        (const zero)
+        (const $ ReqAndOpt HashSet.empty HashSet.empty)
```

Comment:
It would require proving that any two structure chains can be combined into a single one, which sounds like it would be quite difficult. It can be made into an instance of `zero` using this as `zero`, obviously.